File: services/nuldoc/nuldoc-src/pages/SlideListPage.ts
import GlobalFooter from "../components/GlobalFooter.ts";
import GlobalHeader from "../components/SlidesGlobalHeader.ts";
import PageLayout from "../components/PageLayout.ts";
import SlidePageEntry from "../components/SlidePageEntry.ts";
import { Config } from "../config.ts";
import { dateToString } from "../revision.ts";
import { getPostPublishedDate } from "../generators/post.ts";
import { SlidePage } from "../generators/slide.ts";
import { elem, Element, h1, header } from "../dom.ts";

export default async function SlideListPage(
  slides: SlidePage[],
  config: Config,
): Promise<Element> {
  const pageTitle = "スライド一覧";

  return await PageLayout({
    metaCopyrightYear: config.site.copyrightYear,
    metaDescription: "登壇したイベントで使用したスライドの一覧",
    metaTitle: `${pageTitle}｜${config.sites.slides.siteName}`,
    metaAtomFeedHref: `https://${config.sites.slides.fqdn}/slides/atom.xml`,
    site: "slides",
    config,
    children: elem(
      "body",
      { class: "list" },
      GlobalHeader({ config }),
      elem(
        "main",
        { class: "main" },
        header({ class: "page-header" }, h1({}, pageTitle)),
        ...Array.from(slides)
          .sort((s1, s2) => {
            const ta = dateToString(getPostPublishedDate(s1));
            const tb = dateToString(getPostPublishedDate(s2));
            if (ta > tb) return -1;
            if (ta < tb) return 1;
            return 0;
          })
          .map((slide) => SlidePageEntry({ slide, config })),
      ),
      GlobalFooter({ config }),
    ),
  });
}

File: services/nuldoc/lib/nuldoc/pages/slide_list_page.rb
module Nuldoc
  module Pages
    class SlideListPage
      def initialize(slides:, config:)
        @slides = slides
        @config = config
      end

      def render
        config = @config
        page_title = 'スライド一覧'
        sorted = @slides.sort_by { |s| GeneratorUtils.published_date(s) }.reverse

        Components::PageLayout.new(
          meta_copyright_year: config.site.copyright_year,
          meta_description: '登壇したイベントで使用したスライドの一覧',
          meta_title: "#{page_title}｜#{config.sites.slides.site_name}",
          meta_atom_feed_href: "https://#{config.sites.slides.fqdn}/slides/atom.xml",
          site: 'slides',
          config: config,
          children: DOM::HTMLBuilder.new.build do
            body class: 'list' do
              render(Components::SlidesGlobalHeader, config: config)
              main class: 'main' do
                header(class: 'page-header') { h1 { text page_title } }
                sorted.each do |slide|
                  render(Components::SlidePageEntry, slide: slide, config: config)
                end
              end
              render(Components::GlobalFooter, config: config)
            end
          end
        ).build
      end
    end
  end
end
